### ISSUE TITLE
rsyslog general performance - need turbo switch

### ISSUE THREAD

**Issue by kubanowSKY** [rsyslog/rsyslog#4870]
### Expected behavior
queues are dequeued on time

### Actual behavior
queues piles up and rsyslog doesn't process them

### Steps to reproduce the behavior
I've got centralized rsyslog server (16G ram 8 vcpus using `imptcp`) with around 3k servers connected to it and sends huge amounts of logs. Most of them are processed by one ruleset with `FixedArray` queue (for now set at 2kk messages, which for sure isn't enough). In this ruleset I got multiple actions - first is `mmjsonparse` for general and then (using `if`) 7 `omfiles` and 12 `omelasticsearch`. Using `htop` I can see `rsyslogd` threads running at around 50-60% of CPU for around 5s, then around 10s of doing nothing. From `impstats` I can see that cpu resource usage is (5m range, data collected every 30s):
```
{ "name": "parse_logserver2", "origin": "core.action", "processed": 300008, "failed": 0, "suspended": 0, "suspended.duration": 0, "resumed": 0 }
{ "name": "resource-usage", "origin": "impstats", "utime": 943419247, "stime": 402051108, "maxrss": 5578100, "minflt": 65763203, "majflt": 20, "inblock": 184728, "oublock": 20638544, "nvcsw": 7744725, "nivcsw": 143678, "openfiles": 1815 }
{ "name": "logserver2", "origin": "core.queue", "size": 1900028, "enqueued": 300022, "full": 0, "discarded.full": 0, "discarded.nf": 0, "maxqsize": 1900064 }

{ "name": "parse_logserver2", "origin": "core.action", "processed": 384065, "failed": 0, "suspended": 0, "suspended.duration": 0, "resumed": 0 }
{ "name": "resource-usage", "origin": "impstats", "utime": 956624307, "stime": 408229477, "maxrss": 5578100, "minflt": 67126800, "majflt": 20, "inblock": 185680, "oublock": 20991352, "nvcsw": 7879867, "nivcsw": 144811, "openfiles": 1809 }
{ "name": "logserver2", "origin": "core.queue", "size": 1900000, "enqueued": 399972, "full": 0, "discarded.full": 0, "discarded.nf": 0, "maxqsize": 1900064 }

{ "name": "parse_logserver2", "origin": "core.action", "processed": 259816, "failed": 0, "suspended": 0, "suspended.duration": 0, "resumed": 0 }
{ "name": "resource-usage", "origin": "impstats", "utime": 965520386, "stime": 412611308, "maxrss": 5578100, "minflt": 67964038, "majflt": 20, "inblock": 186272, "oublock": 21263488, "nvcsw": 7973765, "nivcsw": 145259, "openfiles": 1810 }
{ "name": "logserver2", "origin": "core.queue", "size": 1887551, "enqueued": 287549, "full": 0, "discarded.full": 0, "discarded.nf": 0, "maxqsize": 1900064 }

{ "name": "parse_logserver2", "origin": "core.action", "processed": 1907871, "failed": 0, "suspended": 0, "suspended.duration": 0, "resumed": 0 }
{ "name": "resource-usage", "origin": "impstats", "utime": 1031811901, "stime": 440273461, "maxrss": 5666504, "minflt": 72637656, "majflt": 20, "inblock": 192448, "oublock": 22812952, "nvcsw": 8543515, "nivcsw": 163073, "openfiles": 1822 }
{ "name": "logserver2", "origin": "core.queue", "size": 1900000, "enqueued": 1862451, "full": 0, "discarded.full": 0, "discarded.nf": 0, "maxqsize": 1900064 }

{ "name": "parse_logserver2", "origin": "core.action", "processed": 600256, "failed": 0, "suspended": 0, "suspended.duration": 0, "resumed": 0 }
{ "name": "resource-usage", "origin": "impstats", "utime": 1052749169, "stime": 447702282, "maxrss": 5686260, "minflt": 73946239, "majflt": 20, "inblock": 193552, "oublock": 23070928, "nvcsw": 8649547, "nivcsw": 166814, "openfiles": 1846 }
{ "name": "logserver2", "origin": "core.queue", "size": 1900010, "enqueued": 600010, "full": 0, "discarded.full": 0, "discarded.nf": 0, "maxqsize": 1900064 }

{ "name": "parse_logserver2", "origin": "core.action", "processed": 600024, "failed": 0, "suspended": 0, "suspended.duration": 0, "resumed": 0 }
{ "name": "resource-usage", "origin": "impstats", "utime": 1074724323, "stime": 460036705, "maxrss": 5710544, "minflt": 75920870, "majflt": 20, "inblock": 194504, "oublock": 23510376, "nvcsw": 8831184, "nivcsw": 168772, "openfiles": 1852 }
{ "name": "logserver2", "origin": "core.queue", "size": 1900014, "enqueued": 600004, "full": 0, "discarded.full": 0, "discarded.nf": 0, "maxqsize": 1900064 }

{ "name": "parse_logserver2", "origin": "core.action", "processed": 218176, "failed": 0, "suspended": 0, "suspended.duration": 0, "resumed": 0 }
{ "name": "resource-usage", "origin": "impstats", "utime": 1083856694, "stime": 464372844, "maxrss": 5755572, "minflt": 76820341, "majflt": 20, "inblock": 194752, "oublock": 23824032, "nvcsw": 8938023, "nivcsw": 169625, "openfiles": 1852 }
{ "name": "logserver2", "origin": "core.queue", "size": 1850000, "enqueued": 199986, "full": 0, "discarded.full": 0, "discarded.nf": 0, "maxqsize": 1900064 }

{ "name": "parse_logserver2", "origin": "core.action", "processed": 231848, "failed": 0, "suspended": 0, "suspended.duration": 0, "resumed": 0 }
{ "name": "resource-usage", "origin": "impstats", "utime": 1090365209, "stime": 466241629, "maxrss": 5784004, "minflt": 77297421, "majflt": 20, "inblock": 195592, "oublock": 23879840, "nvcsw": 8957491, "nivcsw": 169908, "openfiles": 1846 }
{ "name": "logserver2", "origin": "core.queue", "size": 1900010, "enqueued": 250010, "full": 0, "discarded.full": 0, "discarded.nf": 0, "maxqsize": 1900064 }

{ "name": "parse_logserver2", "origin": "core.action", "processed": 132882, "failed": 0, "suspended": 0, "suspended.duration": 0, "resumed": 0 }
{ "name": "resource-usage", "origin": "impstats", "utime": 1094754882, "stime": 468047061, "maxrss": 5784004, "minflt": 77712262, "majflt": 20, "inblock": 195816, "oublock": 24014264, "nvcsw": 8999249, "nivcsw": 170255, "openfiles": 1858 }
{ "name": "logserver2", "origin": "core.queue", "size": 1900000, "enqueued": 149990, "full": 0, "discarded.full": 0, "discarded.nf": 0, "maxqsize": 1900064 }

{ "name": "parse_logserver2", "origin": "core.action", "processed": 717278, "failed": 0, "suspended": 0, "suspended.duration": 0, "resumed": 0 }
{ "name": "resource-usage", "origin": "impstats", "utime": 1119591277, "stime": 478879697, "maxrss": 5784004, "minflt": 79705082, "majflt": 20, "inblock": 197776, "oublock": 24594456, "nvcsw": 9189651, "nivcsw": 171878, "openfiles": 1858 }
{ "name": "logserver2", "origin": "core.queue", "size": 1900006, "enqueued": 700006, "full": 0, "discarded.full": 0, "discarded.nf": 0, "maxqsize": 1900064 }

```
My configs:
```
# template of incoming logs
template(
  name = "m_logserver"
  type = "list"
) {
  constant(value = "<") property(name = "pri") constant(value = ">")
  property(name = "timestamp" dateFormat = "rfc3339") constant(value = " ")
  property(name = "hostname") constant(value = " ")
  property(name = "syslogtag") constant(value = " ")
  constant(value = "@cee:") property(name = "$!")
  constant(value = "\n")
}

# general ruleset
ruleset(
  name = "logserver2"
  queue.type = "fixedArray"
  queue.size = "2000000"

  # watermarks
  queue.fullDelaymark = "2000000"
  queue.discardMark = "2000000"
  queue.lightDelaymark = "1900000"

  # threads
  queue.dequeueBatchSize = "50000"
  queue.minDequeueBatchSize = "10000"
  queue.timeoutShutdown = "1500"
  queue.timeoutEnqueue = "500"
  queue.workerThreads = "8"
  queue.workerThreadMinimumMessages = "250000"

  # disk assisted
  queue.saveOnShutdown = "on"
  queue.filename = "logserver2-unprocessed"
  queue.maxDiskSpace = "4096m"
  queue.maxFileSize = "256m"
)

# actions inside ruleset
# general
action(
  name = "parse_logserver2"
  type = "mmjsonparse"
  action.resumeInterval = "1"
  action.resumeIntervalMax = "30"
  action.reportSuspensionContinuation = "on"
  action.resumeRetryCount = "-1"
)

# omfile with proper if before
action(
  name = "logserver2_omfile_default"
  type = "omfile"
  DynaFile = "f_fromhost_file"
  Template = "m_traditional"
  asyncWriting = "on"
  action.resumeInterval = "1"
  action.resumeIntervalMax = "30"
)

# omelasticsearch with proper if before
action(
  name = "logserver2_omelasticsearch_default"
  type = "omelasticsearch"
  server = "<remote_server>"
  serverport = "9200"
  template = "plain-syslog2"
  searchIndex = "<index>"
  searchType = "_doc"
  dynSearchIndex = "on"
  bulkmode = "on"
  uid = "syslog"
  pwd = "<secret>"
  action.resumeInterval = "30"
  action.resumeRetryCount = "-1"
)
```
Where can I find turbo switch to be able to process 2kk (or even bigger) queue on time?

### Environment
- rsyslog version: rsyslogd  8.1901.0 (aka 2019.01)
- platform: Debian 9 Stretch

**Comment by davidelang** [comment]
in general the number of queues needs to be kept small, only increasing when you 
see a thread hitting 100%. Otherwise you end up with the different threads 
fighting each other for queue access.

if you do a top and then hit H to see the threads, do you see the elasticsearch 
connections eating a lot of cpu? What do the network stats look like (is ES 
accepting messages at full speed, or is it slow to accept messages)

it looks like you are processing messages as quickly as they are arriving, as 
the queue size is not increasing, no matter how many new messages are enqueued.

I would drop/decrease the mindequeuebatchsize. Like the threads parameter, it's 
a 'use only when evreything else fails' type of thing, not something that should 
normally be needed.

David Lang

**Comment by davidelang** [comment]
On Thu, 12 May 2022, Patryk wrote:

>> in general the number of queues needs to be kept small, only increasing when you
> see a thread hitting 100%. Otherwise you end up with the different threads
> fighting each other for queue access.
>
> Are you referring to number of threads? I don't see much performance gain/loss when changing this parameter.

yes, leave the thread count at 1 unless you see the thread maxing out a core 
(100% utilization under top with H)

>> if you do a top and then hit H to see the threads, do you see the elasticsearch
> connections eating a lot of cpu?
>
> I don't see any ES connection threads in `top` nor in `perf`. Only threads 
> from `logserver2` and `impctp`. From `perf` I see `in:imptcp` thread using 
> `imptcp.so` symbol `_init` overhead over 7% over 79k of event `cpu-clock`.

so you really should st threads=1 (or leave it out)

>> What do the network stats look like (is ES
> accepting messages at full speed, or is it slow to accept messages)
>
> For now from my best knowledge there are no issues at ES, and they are have indexing latency between 20ms and 1s.
>
>> it looks like you are processing messages as quickly as they are arriving, as
> the queue size is not increasing, no matter how many new messages are enqueued.
>
> I'm sure I'm discarding a lot of incoming messages. I'm able to create queue 
> with 10kk messages to fill-up in the same amount of time, then performance is 
> still the same. Also because of bigger queue I got delays with delivering logs 
> using `omfile` and `omelasticsearch`. With smaller queue I'm able to deliver 
> logs at time, but I'm sure this server could handle more traffic.

the queue size should not cause delays, it's just the max number of messages 
allowed to be queued. What is bothering me is why you don't seem to drop below 
90000 messages in the queue.

>> I would drop/decrease the mindequeuebatchsize. Like the threads parameter, it's
> a 'use only when evreything else fails' type of thing, not something that should
> normally be needed.
>
> With disabling those:
> ```
> #queue.dequeueBatchSize = "50000"

you want a reasonable batchsize, this is the max number of messages, a few 
hundred to a couple thousand is reasonable

> #queue.minDequeueBatchSize = "10000"
> [...]
> #queue.workerThreads = "8"
> #queue.workerThreadMinimumMessages = "250000"
> ```
> `logserver2` started to use really disk-assisted queue, put around 6kk messages into their queues (2kk in memory, 4kk at disk) and `parse_logserver2` rate decreased 2 times.

pstats data please? and check top again with the new config

David Lang

**Comment by davidelang** [comment]
sorry, queue size of 1900000 almost being the minimum

how frequently are you dumping out the pstats data? pstats doesn't show that 
it's discarding anything, and it never showed the queue as full.

David Lang

**Comment by davidelang** [comment]
On Thu, 12 May 2022, Patryk wrote:

>> What is bothering me is why you don't seem to drop below
> 90000 messages in the queue.
>
> I suppose those messages are just discarded due
> ```
> queue.lightDelaymark = "1900000"
> ```
> as on graphs I see this is "constant" value at `size` of `logserver2` memory queue.
>
> With finally working disk-assisted queue I see discarded messages counter increase. On the other hand this is still value I don't expect.
>
>> you want a reasonable batchsize, this is the max number of messages, a few
> hundred to a couple thousand is reasonable
>
> I thought when increasing `queue.dequeueBatchSize` I could achieve better performance as there are still messages in queue to dequeue.

dequeue batch size is the max number of messages that it will pull from the 
queue to try and process at one time. the default is low (a dozen or so), 
setting it to a few hundred to a few thousand will help. This is the MAX number 
to process, if there are any available to process as it loops through, it will 
process them.

the min number (which you had also set) tells it to delay and not process any 
messages if there aren't that many waiting.

>> pstats data please? and check top again with the new config
>> how frequently are you dumping out the pstats data? pstats doesn't show that
> it's discarding anything, and it never showed the queue as full.
>
> In top I see 3 threads named `rs:logserver2` (`:D` `:R` `[D` <- probably [DA]) - none of them even hit 100% cpu.
> top for rs:logserver every 1s for 60s

that shows that it's not bottlnecked on cpu, so more threads are unlikly to help 
(if you can get a wider display to see what they are saying it would help us 
understand what they are, but the key is that the cpu is not the bottleneck for 
any of them)

> `impstats` are collected every 30s with `resetCounters = "on"`, can change to another interval if you want.
is this the entire set of stats each cycle? or is this just a subset of the 
stats?

> Last 5 minutes:
> ```
>
> { "name": "parse_logserver2", "origin": "core.action", "processed": 262168, "failed": 0, "suspended": 0, "suspended.duration": 0, "resumed": 0 }
> { "name": "resource-usage", "origin": "impstats", "utime": 38648497, "stime": 11099707, "maxrss": 2142956, "minflt": 983504, "majflt": 13, "inblock": 3384, "oublock": 1211464, "nvcsw": 292891, "nivcsw": 1864, "openfiles": 586 }
> { "name": "logserver2[DA]", "origin": "core.queue", "size": 4387534, "enqueued": 9, "full": 1161, "discarded.full": 9, "discarded.nf": 0, "maxqsize": 4423558 }

interesting that the DA queue is discarding messages as full

> { "name": "logserver2", "origin": "core.queue", "size": 1897952, "enqueued": 258048, "full": 0, "discarded.full": 0, "discarded.nf": 0, "maxqsize": 1900008 }

you should have two ES client processes running, one for the main queue, one 
from the DA queue. Can you tell on the ES side what the rate of messages are?

to check if this is a bottleneck in the delivery to ES or if it's in rsyslog, 
try changing the ES output to omfile (or omudp) and validate that it's able to 
process the logs fast enough when it's not delivering to ES.

David Lang

**Comment by davidelang** [comment]
one thing I just noticed in that your omfile is writing to a dynamic filename. 
check the pstats data for that and see if it's reporting a large number of cache 
misses. If so you need to increase the dynafilecachesize to something larger 
than the working set of files you will be writing to.

10-12k messages/sec is very low for just writing to files, but if you are 
thrashing the file cache, that would easily explain it.

David Lang


  On Thu, 12 May 2022, 
Patryk wrote:

> Date: Thu, 12 May 2022 14:57:09 -0700
> From: Patryk ***@***.***>
> Reply-To: rsyslog/rsyslog
>     ***@***.***>
> To: rsyslog/rsyslog ***@***.***>
> Cc: David Lang ***@***.***>, Comment ***@***.***>
> Subject: Re: [rsyslog/rsyslog] rsyslog general performance - need turbo switch
>      (Issue #4870)
> 
>> that shows that it's not bottlnecked on cpu, so more threads are unlikly to help
> (if you can get a wider display to see what they are saying it would help us
> understand what they are, but the key is that the cpu is not the bottleneck for
> any of them)
>
> It's full name of the thread.
> ```
> /proc/21107/task/21130# cat status
> Name:	rs:logserver2:R
> ```
>
>> is this the entire set of stats each cycle? or is this just a subset of the
> stats?
>
> Subset of stats.
>
>> you should have two ES client processes running, one for the main queue, one
> from the DA queue. Can you tell on the ES side what the rate of messages are?
>
> I don't see any separate thread about `omelasticsearch`. According to `impstats` `omelasticsearch` process around 1200 msg/s for now. Before changes it was 2600 msg/s.
>
>> to check if this is a bottleneck in the delivery to ES or if it's in rsyslog,
> try changing the ES output to omfile (or omudp) and validate that it's able to
> process the logs fast enough when it's not delivering to ES.
>
> I tried disabling ES lately - there was no performance gain on `parse_logserver2` nor any `omfile`.
> Tried do the same now - `parse_logserver2` ratio increased - from 10k msg/s to 12k msg/s - bringing back `omelasticsearch`.
> Setting `queue.dequeueBatchSize` to 10000 gave the same result as disabling `omelasticsearch`.
>
>

**Comment by davidelang** [comment]
so look for dynafile lines in the pstats output (ideally post 1-2 cycles of full 
output) I'll bet that you have a very high number of cache miss items when you 
have omfile enabled. increase dynafilecachesize until it's larger than the 
number of files you expect to be writing to and you will see that drop 
drastically and the throughput writing to files should increase as well.

At that point look to see if you have a thread maxing out the cpu, and if so, 
increase the number of threads allowed.

David Lang

  On Thu, 
12 May 2022, Patryk wrote:

**Comment by davidelang** [comment]
I don't think this is the issue, but asyncwrite only really helps when you are 
doing things like gzipping the output file, with simple files it adds more 
overhead (and lots of threads)

in terms of dynafile cache, how many files are you normally creating at a time?

it seems to me that the problem is probably in the omelasticsearch module, but 
I'm not familiar with it's innards, so I'm nibbling around the edges at the 
things that I can know about.

it may be worth putting a queue on the omelasticsearch actions, to decouple them 
from each other and let you see the behavior of each of them  (and to let them 
all run in parallel). Since this significantly changes the way they will work in 
parallel, I would go back and reduce the thread count again when you test that.

David Lang

**Comment by davidelang** [comment]
if you do a strace on the thread that's writing things out, if you are way under 
the number of dynamic files that you need, you will see that every write has to 
close and open a file, that is extremely slow

when you have async writing enabled (or multiple threads) you have locks that 
need to be acquired for each log line.

so disable async writing, set threads to 1, disable the ES output and set the 
dynafilecachesize large enough that you don't have a huge number of cache misses

you will see that the traffic rate can then be very high (hundreds of thousands 
of messages/sec)

then we can add back in the ES output and deal with the issues that that adds.

David Lang

**Comment by rgerhards** [comment]
> with the default dynafilecachesize, if you are getting a 99%+ hit rate, you
> aren't using very many files, are you sure about that?
>

This can be possible, even with very many files. It's when a very
small set of files receives almost all messages. But best to
double-check.

I also have some troubleshooting ideas, but let's do one step after another.

Rainer

**Comment by rgerhards** [comment]
Apologies for the very late reply. This is probably no longer relevant to the original reporter, but I am adding the conclusion for people who arrive here via search while we work through the backlog. We do not close rsyslog issues lightly, and we are not closing this merely because of age.

Triage result: valid non-code answer.

Reasoning: This was a long performance-tuning support thread. The discussion covered worker threads, batch sizing, dynafile cache, async writing, and output bottlenecks; no specific current defect was left open.

If this still reproduces on a current rsyslog release and maintained platform, please open a new issue with the current version, complete configuration, logs, and a minimal reproducer so we can investigate it as a fresh current problem.
